전남대 BE_강명덕 4주차 과제 (3단계)

Profile-exe opened the pull request:
## 구현 내용
- 옵션 재고 수량을 빼는 기능을 추가했습니다.

## 코드를 작성하며 어려웠던 점
요구사항에 따라 옵션 재고 수량 기능에 대한 테스트코드를 구현하기 전에 상품 도메인의 서비스 레이어 테스트를 구현해보고 싶었습니다. 

`Mockito`를 이용해서 상품 서비스 테스트를 구현해보면 옵션 서비스는 쉬울 것이라 판단했습니다. 다만 상품 서비스의 메서드가 많고 처음 접하는 테스트 라이브러리를 이용하다보니 익숙하지 않아 시간이 너무 많이 걸렸습니다.

## 코드 리뷰 시, 멘토님이 중점적으로 리뷰해줬으면 하는 부분
상품 서비스, 옵션 서비스에 대한 테스트코드를 작성했는데, 적절하게 작성되었는지 리뷰해주시면 감사하겠습니다!

Profile-exe committed:
docs(README.md): Step3 요구사항 작성

Profile-exe committed:
feat(OptionRepository): 옵션 리파지토리 구현

Profile-exe committed:
feat(Option): 옵션에 수량 차감을 위한 비즈니스 메서드 구현

Profile-exe committed:
feat: 수량 차감이 불가능한 경우에 대한 비즈니스 예외 구현

Profile-exe committed:
feat(OptionService): 옵션 서비스에 수량 차감 비즈니스 메서드 구현

Profile-exe committed:
fix(ProductReqDto): 옵션 검증을 위한 Valid 어노테이션 위치 조정

Profile-exe committed:
test(ProductE2ETest): 옵션 검증에 대한 테스트코드 구현

Profile-exe committed:
feat(ProductService): InvalidDataAccessApiUsageException 예외처리로 변경

Profile-exe committed:
test(ProductServiceTest): 상품 서비스 테스트 구현

Profile-exe committed:
test(OptionServiceTest): 옵션 서비스 테스트 구현

hocaron commented on src/main/java/gift/option/service/OptionService.java:
```diff
@@ -28,6 +36,17 @@ public void updateOptions(Product product, List<OptionReqDto> optionReqDtos) {
+                if (option.getQuantity() < quantity) {
+                    throw OptionNotEnoughStockException.EXCEPTION;
+                }
+                option.subtractStock(quantity);
```
Stcok 이라는 단어도 좋지만, 현재 코드에서 사용하는 subtractQuantity 또는 subtract 그 자체로도 의미를 표현할 수 있을 것 같아요😀

Profile-exe replied:
메서드명을 고민했었는데 역시 바꾸는게 좋을 것 같네요!

[63be7a3](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-enhancement/pull/359/commits/63be7a343ad1e28ba47fd824930e228d48f99d50) 커밋에 반영했습니다!

hocaron reviewed (approved):
명덕님, 안녕하세요👋

필요한 필드만 mocking 해주셨네요👍
pr 머지까지 기간이 조금 남아서 테스트 시에 사용되는 객체를 재사용할 수 있도록 리팩토링해보는 것도 좋을 것 같은데, 어떠실까요~?!!

hocaron commented:
객체를 관리하는 부분은 static 으로 만들어서 재사용하면 좋을 것 같아요. 이런 객체를 test fixture 라고 부르는데요, 멀티 모듈인 경우에는 다른 모듈에서 정의한 객체도 재사용가능해서 많이 사용하고 있어요😀 혹시 적용가능한 부분이 있다면 new 로 새로운 객체를 만드는 것 대신 static 으로 선언하여 재사용이 가능할 것 같아요.

https://bottom-to-top.tistory.com/58

Profile-exe committed:
refactor(Option): subtractQuantity로 메서드명 변경

Profile-exe committed:
chore(build.gradle): java-test-fixtures 플러그인 적용

Profile-exe commented:
> 객체를 관리하는 부분은 static 으로 만들어서 재사용하면 좋을 것 같아요. 이런 객체를 test fixture 라고 부르는데요, 멀티 모듈인 경우에는 다른 모듈에서 정의한 객체도 재사용가능해서 많이 사용하고 있어요😀 혹시 적용가능한 부분이 있다면 new 로 새로운 객체를 만드는 것 대신 static 으로 선언하여 재사용이 가능할 것 같아요.
> 
> https://bottom-to-top.tistory.com/58

이 부분을 보고 [공식문서](https://docs.gradle.org/current/userguide/java_testing.html#sec:java_test_fixtures)를 포함해 구글링을 하며 적용해보려 하는데 적용이 안됩니다...

`test`내부 클래스들에서 `testFixture`의 클래스를 참조하지 못합니다.

<br>

그래들 의존성에도 잘 들어가있고

<details>
<summary> gradle 의존성 확인 이미지  </summary>

![image](https://github.com/user-attachments/assets/c32bffe3-fefb-433b-a3ac-4fdc0a9d4b56)

</details>

<br>

`build.gradle`에도 의존성을 추가했습니다. 혹시 몰라 `bootJar`, `jar` 코드까지 함께 넣어서 빌드 해보기도 했습니다.

<details>
<summary> build.gradle 코드 </summary>

```gradle
plugins {
    id 'java'
    id 'org.springframework.boot' version '3.3.1'
    id 'io.spring.dependency-management' version '1.1.5'

    // A Java Library
    id 'java-library'
    // which produces test fixtures
    id 'java-test-fixtures'
    // and is published
    id 'maven-publish'
}

bootJar { enabled = false }
jar { enabled = true }

group = 'camp.nextstep.edu'
version = '0.0.1-SNAPSHOT'

java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(21)
    }
}

repositories {
    mavenCentral()
}

sourceSets {
    testFixtures {
        java {
            srcDir file('src/testFixtures/java')
        }
        resources {
            srcDir file('src/testFixtures/resources')
        }
    }
}

dependencies {
    implementation 'org.springframework.boot:spring-boot-starter-jdbc'
    implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
    implementation 'org.springframework.boot:spring-boot-starter-web'
    implementation 'org.springframework.boot:spring-boot-starter-validation'
    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'

    runtimeOnly 'com.h2database:h2'

    compileOnly 'io.jsonwebtoken:jjwt-api:0.12.6'
    runtimeOnly 'io.jsonwebtoken:jjwt-impl:0.12.6'
    runtimeOnly 'io.jsonwebtoken:jjwt-jackson:0.12.6'

    testImplementation 'org.springframework.boot:spring-boot-starter-test'

    testFixturesImplementation 'org.springframework.boot:spring-boot-starter-test'
    testRuntimeOnly 'org.junit.platform:junit-platform-launcher'
}

tasks.named('test') {
    useJUnitPlatform()
}
```
</details>

<br>

클래스의 경우 빌드도 잘 되었습니다.

<details>

<summary> 빌드된 코드 </summary>

![image](https://github.com/user-attachments/assets/72bcdde4-bf89-4d1e-89a5-a8bfd762b4a8)

</details>

<br>

사용하려는 `ProductFixture` 클래스 또한 `public` 접근지정자를 적용했습니다.

<details>

<summary> ProductFixture 코드 </summary>

```java
package gift.product;

import gift.category.entity.Category;
import gift.product.entity.Product;

public class ProductFixture {

    public static Product createProduct(String name, Integer price, String imageUrl, Category category) {
        return new Product(name, price, imageUrl, category);
    }

    public static Product createProduct(String name, Integer price, String imageUrl) {
        return new Product(name, price, imageUrl, null);
    }

    public static Product createProduct() {
        return new Product("name", 1000, "imageUrl", null);
    }

    public static Product createProduct(Category category) {
        return new Product("name", 1000, "imageUrl", category);
    }
}
```

</details>

다른 모듈이 아닌 동일한 모듈에서 `testFixture`를 적용한 것인데 `test`에서 왜 접근이 안되는지, IDE 문제인 것인지 잘 모르겠습니다.

[저랑 비슷한 상황](https://discuss.gradle.org/t/intellij-tests-not-compiling-due-to-test-fixtures/41837/9)의 글을 보았는데, 

딱히 해결법이 없는 것 같아서 `Settings`에서 `Build Tools` -> `Gradle`의 빌드 도구도 `Gradle` 빌드해보기 -> `Intellij IDEA`로 설정 후 빌드해보기 -> `Gradle`로 다시 빌드해보기도 여전히 `test`에서 `testFixture`에 대해 참조가 불가능했습니다.

`Gradle 8.5` 버전이 Java 21을 지원한대서 `8.4`를 `8.5`로 변경해보아도 여전히 참조가 불가능했습니다...

<br>

`test` 내에서 `fixture` 클래스들을 만들고 사용할 수 있지만 이런 간단한 플러그인 적용도 안되는 상황이 너무 답답해서 계속 시도하다가 답변 남깁니다

hocaron commented:
역시 고민 많이 해주셨네요👍
명덕님 혹시 gift 폴더를 java 밑에 넣어보고, ProductFixture 잘 가져와지는지 확인 부탁드릴 수 있을까요?!

<img width="459" alt="image" src="https://github.com/user-attachments/assets/70408394-fd30-4120-baa5-5a7f97ce1422">

엇 지금 보니까 잘 해주신 것 같네요?! 명덕님이 공유주신 build.gradle 설정 및 fixture class 만 생성했을 때, 테스트 클래스에서 fixture 잘 들고 오는 상태라, 혹시 invalid cache 한번 날리고 다시 시도해보실 수 있을까요?!

<img width="778" alt="image" src="https://github.com/user-attachments/assets/1dfb70ae-9b16-4f01-9965-1190fc30ea0c">

Profile-exe commented:
> 엇 지금 보니까 잘 해주신 것 같네요?! 명덕님이 공유주신 build.gradle 설정 및 fixture class 만 생성했을 때, 테스트 클래스에서 fixture 잘 들고 오는 상태라, 혹시 invalid cache 한번 날리고 다시 시도해보실 수 있을까요?!

와 `invalid cache`로 `file system cache, local history` 지우니까 잘 되네요 정말 감사합니다! 적용은 내일까지 해보겠습니다!

Profile-exe committed:
chore(Gradle): Java 21을 지원하는 8.5 버전으로 변경

Profile-exe committed:
feat(RestPage): 오류 방지를 위한 기본 생성자 정의

오류메시지: Cannot construct instance of `gift.utils.RestPage` (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator

Profile-exe committed:
test: Test Fixtures 생성

회원, 상품, 위시 리스트, 카테고리, 옵션에 대한 test fixture 클래스들 생성

Profile-exe committed:
test: Test Fixtures 적용

Profile-exe committed:
test(WishListE2ETest): 테스트 종료 후 카테고리 데이터 제거

카테고리 E2E 테스트에 영향을 주는 상황이 발생해 해당 카테고리들을 모두 제거

Profile-exe commented:
Test Fixture 적용했습니다!

확실히 코드도 깔끔해지고 불필요한 파라미터를 넣느라 코드가 길어지지 않아 좋았습니다.
이렇게 좋은 개념을 알려주셔서 감사합니다

처음에 이를 적용하며 [공식문서](https://docs.gradle.org/current/userguide/java_testing.html#sec:java_test_fixtures)의 예시처럼 자주 사용되는 값들을 가진 객체들에 대해 이름을 부여하고 사용하려 했습니다.

```java
public class Simpsons {
    private static final Person HOMER = new Person("Homer", "Simpson");
    private static final Person MARGE = new Person("Marjorie", "Simpson");
    private static final Person BART = new Person("Bartholomew", "Simpson");
    private static final Person LISA = new Person("Elisabeth Marie", "Simpson");
    private static final Person MAGGIE = new Person("Margaret Eve", "Simpson");
    private static final List<Person> FAMILY = new ArrayList<Person>() {{
        add(HOMER);
        add(MARGE);
        add(BART);
        add(LISA);
        add(MAGGIE);
    }};

    public static Person homer() { return HOMER; }

    public static Person marge() { return MARGE; }

    public static Person bart() { return BART; }

    public static Person lisa() { return LISA; }

    public static Person maggie() { return MAGGIE; }

    // ...
```


하지만 식별자로 해당 객체의 값들을 감추어 기대값을 해당 테스트케이스에서 바로 확인할 수 없다고 생각해 이런 방법은 사용하지 않기로 했습니다.

만약 매개변수를 하나만 갖는 클래스들은 위와 같은 방법이 유용할 수 있다고 생각합니다.
이번주도 꼼꼼히 리뷰해 주셔서 정말 감사드립니다.

hocaron commented:
명덕님 4주차 고생 많으셨습니다👍👍👍

다음 주차에서 뵙기위해 해당 PR은 머지하겠습니다🙇‍♀️
